text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1309 @@
+
+### Native class syntax
+
+One of the benefits of native class syntax is that it is used outside of Ember,
+so as time goes on we will be able to assume there is more general knowledge of
+it, and provide links to documentation for it for users who are not familiar.
+During this transitionary period though we should add a more thorough primer of
+the syntax to our guides, and explicitly call out the differences between
+classic class syntax and native class syntax, including:
+
+* Usage of `constructor` in classes which _do not_ extend from classic classes.
+  Otherwise, always use `init`. This will be tricky, because even when using
+  native classes to extend from classic classes, you should still use `init`.
+
+* Nuances of class fields - they run _after_ `super`, and _before_ user code.
+
+* Benefits of class field initializers, and how they can be used to do much of
+  the work that would otherwise be done in the `constructor` or `init`
+
+* Expense of class fields - new objects and functions are created for every
+  instance, so users should also be careful with them.
+
+### Arguments
+
+For arguments, namespacing makes sense in general as an API choice and is common
+in other frameworks (`props` in React, etc.). We should be sure to cover this
+thoroughly for users who are used to classic components, but it shouldn't
+require too much explaining.
+
+Immutability will be a bigger sticking point in general, in particular the
+inability to provide default argument values. This is easy enough to work around
+using an `{{or}}` helper in templates:
+
+```hbs
+Hello, {{or @firstName "friend"}}!
+```
+
+Or a defaulting alias decorator in the class:
+
+```hbs
+Hello, {{this.firstName}}!
+```
+```js
+export default class Greeting extends GlimmerComponent {
+  @tracked
+  get firstName() {
+    return this.args.firstName || 'friend';
+  }
+}
+```
+
+But does add a bit of boilerplate to components. Users will also have to be
+careful when attempting to override these "defaults" in subclasses, since it is
+not as simple as overriding a class field or property. We should guide users to
+use template-only components to "curry" components when trying to provide
+defaults in subclasses instead, leveraging the outer HTML semantics of
+Glimmer components:
+
+```hbs
+<!-- components/button.hbs -->
+<button class="button {{@type}}">
+  <i class="icon {{@type}}"></i>
+  {{yield}}
+</button>
+```
+```hbs
+<!-- component/success-button.hbs -->
+<Button @type="success">{{yield}}</Button>
+```
+```hbs
+<!-- component/danger-button.hbs -->
+<Button @type="danger">{{yield}}</Button>
+```
+
+#### Lifecycle hooks and properties
+
+For new users, and users of other frameworks, lifecycle hooks will be very
+minimal and fairly easy to understand. The lack of render hooks may be the more
+difficult part to understand, and we'll have to lean on the documentation for
+element modifiers and make sure that is _really_ excellent to get the concepts
+there across.
```

Comment:
Actually, I think that native custom elements really capture this perfectly. They have _no sub components_, and they are effectively _real DOM elements_. They are aligned with the browser, and so are modifiers.

Modifiers are running code by directly connecting it to where it is concerned, instead of through some arbitrary boundary point. We are ensuring correctness of code through correctness of _form_ in the template language.

In the future, we also plan on allowing modifiers to work with components, so users will have more flexibility with them:

```hbs
<Button {{did-render this.setupButton}}></Button>
```

If we accept that modifiers _will_ be a major concept in Ember, then I think it makes perfect sense that they also become _the_ concept which allows users to interact with the DOM directly.